File: elevaite/python_apps/agent_studio/agent-studio/db/models/tools.py

import uuid
from datetime import datetime
from typing import List, Optional

from sqlalchemy import (
    Boolean,
    DateTime,
    ForeignKey,
    Integer,
    String,
    UniqueConstraint,
    Text,
)
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from .base import Base, get_utc_datetime

class ToolCategory(Base):
    __tablename__ = "tool_categories"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    category_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), unique=True, nullable=False, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String, nullable=False, unique=True)
    description: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    icon: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    color: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=get_utc_datetime)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=get_utc_datetime, onupdate=get_utc_datetime)

    tools = relationship("Tool", back_populates="category")

class MCPServer(Base):
    __tablename__ = "mcp_servers"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    server_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), unique=True, nullable=False, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String, nullable=False)
    description: Mapped[Optional[str]] = mapped_column(String, nullable=True)

    host: Mapped[str] = mapped_column(String, nullable=False)
    port: Mapped[int] = mapped_column(Integer, nullable=False)
    protocol: Mapped[str] = mapped_column(String, default="http")  # http, https, ws, wss
    endpoint: Mapped[Optional[str]] = mapped_column(String, nullable=True)  # API endpoint path

    auth_type: Mapped[Optional[str]] = mapped_column(String, nullable=True)  # none, bearer, basic, api_key
    auth_config: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)  # Auth configuration

    status: Mapped[str] = mapped_column(String, default="active")  # active, inactive, error, maintenance
    last_health_check: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    health_check_interval: Mapped[int] = mapped_column(Integer, default=300)  # seconds
    consecutive_failures: Mapped[int] = mapped_column(Integer, default=0)

    version: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    capabilities: Mapped[Optional[List[str]]] = mapped_column(JSONB, nullable=True)
    tags: Mapped[Optional[List[str]]] = mapped_column(JSONB, nullable=True)

    registered_at: Mapped[datetime] = mapped_column(DateTime, default=get_utc_datetime)
    last_seen: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=get_utc_datetime, onupdate=get_utc_datetime)

    tools = relationship("Tool", back_populates="mcp_server")

    __table_args__ = (UniqueConstraint("name", name="uix_mcp_server_name"),)

class Tool(Base):
    __tablename__ = "tools"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    tool_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), unique=True, nullable=False, default=uuid.uuid4)

    name: Mapped[str] = mapped_column(String, nullable=False)
    display_name: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    description: Mapped[str] = mapped_column(String, nullable=False)
    version: Mapped[str] = mapped_column(String, default="1.0.0")

    tool_type: Mapped[str] = mapped_column(String, nullable=False)  # local, remote, mcp
    execution_type: Mapped[str] = mapped_column(String, default="function")  # function, api, command

    parameters_schema: Mapped[dict] = mapped_column(JSONB, nullable=False)  # OpenAI function schema
    return_schema: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)  # Expected return format

    module_path: Mapped[Optional[str]] = mapped_column(String, nullable=True)  # Python module path
    function_name: Mapped[Optional[str]] = mapped_column(String, nullable=True)  # Function name

    mcp_server_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("mcp_servers.server_id"), nullable=True
    )
    remote_name: Mapped[Optional[str]] = mapped_column(String, nullable=True)  # Name on the remote server

    api_endpoint: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    http_method: Mapped[Optional[str]] = mapped_column(String, nullable=True)  # GET, POST, PUT, DELETE
    headers: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)
    auth_required: Mapped[bool] = mapped_column(Boolean, default=False)

    category_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("tool_categories.category_id"), nullable=True
    )
    tags: Mapped[Optional[List[str]]] = mapped_column(JSONB, nullable=True)

    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    is_available: Mapped[bool] = mapped_column(Boolean, default=True)  # Runtime availability
    last_used: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    usage_count: Mapped[int] = mapped_column(Integer, default=0)

    documentation: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    examples: Mapped[Optional[List[dict]]] = mapped_column(JSONB, nullable=True)

    created_at: Mapped[datetime] = mapped_column(DateTime, default=get_utc_datetime)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=get_utc_datetime, onupdate=get_utc_datetime)

    category = relationship("ToolCategory", back_populates="tools")
    mcp_server = relationship("MCPServer", back_populates="tools")

    __table_args__ = (UniqueConstraint("name", name="uix_tool_name"),)

File: elevaite/apps/command_agent_version1/app/lib/discriminators/tools.ts
import type {
  Tool,
  ToolCategory,
  MCPServer,
  ToolExecutionResponse,
} from "../interfaces/tools";
import { isObject } from "./common";

export function isTool(data: unknown): data is Tool {
  return isToolObject(data);
}

export function isToolArray(data: unknown): data is Tool[] {
  return Array.isArray(data) && data.every((item) => isToolObject(item));
}

export function isToolCategory(data: unknown): data is ToolCategory {
  return isToolCategoryObject(data);
}

export function isToolCategoryArray(data: unknown): data is ToolCategory[] {
  return (
    Array.isArray(data) && data.every((item) => isToolCategoryObject(item))
  );
}

export function isMCPServer(data: unknown): data is MCPServer {
  return isMCPServerObject(data);
}

export function isMCPServerArray(data: unknown): data is MCPServer[] {
  return Array.isArray(data) && data.every((item) => isMCPServerObject(item));
}

export function isToolExecutionResponse(
  data: unknown
): data is ToolExecutionResponse {
  return isToolExecutionResponseObject(data);
}

function isToolObject(item: unknown): item is Tool {
  return (
    isObject(item) &&
    "id" in item &&
    "tool_id" in item &&
    "name" in item &&
    "description" in item &&
    "version" in item &&
    "tool_type" in item &&
    "execution_type" in item &&
    "parameters_schema" in item &&
    "requires_auth" in item &&
    "timeout_seconds" in item &&
    "retry_count" in item &&
    "is_active" in item &&
    "is_available" in item &&
    "created_at" in item &&
    "updated_at" in item &&
    "usage_count" in item &&
    "success_count" in item &&
    "error_count" in item &&
    typeof item.id === "number" &&
    typeof item.tool_id === "string" &&
    typeof item.name === "string" &&
    typeof item.description === "string" &&
    typeof item.version === "string" &&
    ["local", "remote", "mcp"].includes(item.tool_type as string) &&
    typeof item.execution_type === "string" &&
    typeof item.parameters_schema === "object" &&
    typeof item.requires_auth === "boolean" &&
    typeof item.timeout_seconds === "number" &&
    typeof item.retry_count === "number" &&
    typeof item.is_active === "boolean" &&
    typeof item.is_available === "boolean" &&
    typeof item.created_at === "string" &&
    typeof item.updated_at === "string" &&
    typeof item.usage_count === "number" &&
    typeof item.success_count === "number" &&
    typeof item.error_count === "number" &&
    (!("display_name" in item) ||
      item.display_name === null ||
      typeof item.display_name === "string") &&
    (!("return_schema" in item) ||
      item.return_schema === null ||
      typeof item.return_schema === "object") &&
    (!("module_path" in item) ||
      item.module_path === null ||
      typeof item.module_path === "string") &&
    (!("function_name" in item) ||
      item.function_name === null ||
      typeof item.function_name === "string") &&
    (!("remote_name" in item) ||
      item.remote_name === null ||
      typeof item.remote_name === "string") &&
    (!("rate_limit_per_minute" in item) ||
      item.rate_limit_per_minute === null ||
      typeof item.rate_limit_per_minute === "number") &&
    (!("category_id" in item) ||
      item.category_id === null ||
      typeof item.category_id === "string") &&
    (!("mcp_server_id" in item) ||
      item.mcp_server_id === null ||
      typeof item.mcp_server_id === "string") &&
    (!("created_by" in item) ||
      item.created_by === null ||
      typeof item.created_by === "string") &&
    (!("tags" in item) || item.tags === null || Array.isArray(item.tags)) &&
    (!("last_used" in item) ||
      item.last_used === null ||
      typeof item.last_used === "string") &&
    (!("average_execution_time_ms" in item) ||
      item.average_execution_time_ms === null ||
      typeof item.average_execution_time_ms === "number") &&
    (!("category" in item) ||
      item.category === null ||
      isToolCategoryObject(item.category)) &&
    (!("mcp_server" in item) ||
      item.mcp_server === null ||
      isMCPServerObject(item.mcp_server))
  );
}

function isToolCategoryObject(item: unknown): item is ToolCategory {
  return (
    isObject(item) &&
    "id" in item &&
    "category_id" in item &&
    "name" in item &&
    "created_at" in item &&
    "updated_at" in item &&
    typeof item.id === "number" &&
    typeof item.category_id === "string" &&
    typeof item.name === "string" &&
    typeof item.created_at === "string" &&
    typeof item.updated_at === "string" &&
    (!("description" in item) ||
      item.description === null ||
      typeof item.description === "string") &&
    (!("icon" in item) ||
      item.icon === null ||
      typeof item.icon === "string") &&
    (!("color" in item) ||
      item.color === null ||
      typeof item.color === "string")
  );
}

function isMCPServerObject(item: unknown): item is MCPServer {
  return (
    isObject(item) &&
    "id" in item &&
    "server_id" in item &&
    "name" in item &&
    "host" in item &&
    "port" in item &&
    "protocol" in item &&
    "status" in item &&
    "health_check_interval" in item &&
    "consecutive_failures" in item &&
    "registered_at" in item &&
    "updated_at" in item &&
    typeof item.id === "number" &&
    typeof item.server_id === "string" &&
    typeof item.name === "string" &&
    typeof item.host === "string" &&
    typeof item.port === "number" &&
    typeof item.protocol === "string" &&
    typeof item.status === "string" &&
    typeof item.health_check_interval === "number" &&
    typeof item.consecutive_failures === "number" &&
    typeof item.registered_at === "string" &&
    typeof item.updated_at === "string" &&
    (!("description" in item) ||
      item.description === null ||
      typeof item.description === "string") &&
    (!("endpoint" in item) ||
      item.endpoint === null ||
      typeof item.endpoint === "string") &&
    (!("auth_type" in item) ||
      item.auth_type === null ||
      typeof item.auth_type === "string") &&
    (!("auth_config" in item) ||
      item.auth_config === null ||
      typeof item.auth_config === "object") &&
    (!("version" in item) ||
      item.version === null ||
      typeof item.version === "string") &&
    (!("capabilities" in item) ||
      item.capabilities === null ||
      Array.isArray(item.capabilities)) &&
    (!("tags" in item) || item.tags === null || Array.isArray(item.tags)) &&
    (!("last_health_check" in item) ||
      item.last_health_check === null ||
      typeof item.last_health_check === "string") &&
    (!("last_seen" in item) ||
      item.last_seen === null ||
      typeof item.last_seen === "string")
  );
}

function isToolExecutionResponseObject(
  item: unknown
): item is ToolExecutionResponse {
  return (
    isObject(item) &&
    "status" in item &&
    "execution_time_ms" in item &&
    "tool_id" in item &&
    "timestamp" in item &&
    typeof item.status === "string" &&
    ["success", "error", "timeout"].includes(item.status) &&
    typeof item.execution_time_ms === "number" &&
    typeof item.tool_id === "string" &&
    typeof item.timestamp === "string" &&
    (!("result" in item) || item.result !== undefined) &&
    (!("error_message" in item) ||
      item.error_message === null ||
      typeof item.error_message === "string")
  );
}
